KDSoapUdpClient

caspermeijn opened the pull request:
KDSoapUdpClient provides an interface for implementing a
SOAP-over-UDP[1] client.

It includes tests and documentation. I tested it with (kdsoap-ws-discovery-client)[https://gitlab.com/caspermeijn/kdsoap-ws-discovery-client].

Fixes #175 

[1] https://docs.oasis-open.org/ws-dd/soapoverudp/1.1/os/wsdd-soapoverudp-1.1-spec-os.html

caspermeijn commented:
@apachelogger Can you have a look at this?

apachelogger commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,111 @@
+void KDSoapUdpClientPrivate::readyReadIPv4()
+{
+    while (socketIPv4->hasPendingDatagrams()) {
+        QNetworkDatagram datagram = socketIPv4->receiveDatagram();
```
The two lines could be merged.
In fact, considering this and ipv6 are the same code except for the socket you could just pass the socket in, or cast sender().

caspermeijn replied:
I changed that code, also for Qt4 compatibility.

apachelogger commented on src/KDSoapClient/KDSoapUdpClient.h:
```diff
@@ -0,0 +1,126 @@
+     * \see bind()
+     * \since 1.9
+     */
+    void receivedMessage(const KDSoapMessage& message, const KDSoapHeaders& headers, const QHostAddress& address, quint16 port);
```
& on wrong side of space

apachelogger commented on src/KDSoapClient/KDSoapUdpClient.h:
```diff
@@ -0,0 +1,126 @@
+     * \param port The UDP port to send the message to
+     * \since 1.9
+     */    
+    bool sendMessage(const KDSoapMessage &message, const KDSoapHeaders &headers, const QHostAddress& address, quint16 port);
```
& on wrong side of space

apachelogger commented on src/KDSoapClient/KDSoapUdpClient_p.h:
```diff
@@ -0,0 +1,41 @@
+class KDSoapUdpClientPrivate
+{
+public:
+    KDSoapUdpClientPrivate(KDSoapUdpClient * q) : q_ptr(q){;}
```
unnecessary ;

apachelogger commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,111 @@
+bool KDSoapUdpClient::bind(quint16 port) {
+    Q_D(KDSoapUdpClient);
+    bool rc = true;
+    rc = d->socketIPv4->bind(QHostAddress::AnyIPv4, port, QUdpSocket::ShareAddress) && rc;
```
Is there a particular reason the ports need to be shared?

caspermeijn replied:
I was under the impression that other program could use the port as well. This would be useful if multiple programs want to do WS-Discovery.

apachelogger replied:
Only for the server though. Clients have arbitrary ports, no? E.g. I see windows10 send Probes from all sorts of ports, but the destination is always 3702.

caspermeijn replied:
You are right, this appear to work, I will allow both uses then. For some reason my firewall will refuse responses to the random port :-(

dfaure-kdab commented:
>     * I think kdsoap is still meant to build with Qt4, so you cannot use the lambda based connect syntax. Oh and I think you need qt includes with their directory so <QtCore/QObject> etc.

No, <QObject> works with both Qt4 and Qt5.

>     * I would also use `&=` and `||=` instead of `= foo & bar` respectively `= foo || bar`. The former are distinctly more readable IMHO.

`||=` does not exist in C++, and `&=` would do the wrong thing for boolean expressions. Please stick to `foo = foo && bar` and `foo = foo || bar`, in an order that matches whether you want short circuit evaluation or not.

caspermeijn commented:
>     * I think kdsoap is still meant to build with Qt4, so you cannot use the lambda based connect syntax. Oh and I think you need qt includes with their directory so <QtCore/QObject> etc.

I made the code compatible with Qt4. However I could only compile test it.

> 
>     * Still some disabled qdebugs littered about.

This appears the way to add optional debug messages to KDSoap.

> 
>     * I would also use `&=` and `||=` instead of `= foo & bar` respectively `= foo || bar`. The former are distinctly more readable IMHO.
> 
>     * For the tests I'd bind port=0 to dynamically get a free port.

I was able to do this for the sending test. However I have to add extra code to make is happens receiving code. Do you think that is worth the trouble?

> 
>     * I am not sure how consistent kdsoap's coding style is but even within the diff the use of spaces is inconsistent. I'd suggest putting a space after every if `if (`, as well as a space before * and & following a type `const QHostAddress &`.

I think I fixed most of those.

> 
> 
> On the not nit-picky front: Since this client is meant to enable ws-discovery implementations, the current code doesn't look quite sufficient for [ws-discovery](http://specs.xmlsoap.org/ws/2005/04/discovery/ws-discovery.pdf)'s needs. Specifically:
> 
> > Messages sent over UDP MUST be sent using SOAP over UDP [SOAP/UDP]. To compensate for possible UDP unreliability, senders MUST use the example transmission algorithm in Appendix I of SOAP over UDP.
> 
> Now, while discovery built on top of the client could do that, I think it stands to reason that it's more expected and efficient if we do it in the client. In particular since all request-response uses of this client will suffer from the inherit unreliability of UDP, so most uses need retransmission to overcome UDP's limitations.

This code doesn't implement a full request-response client, as it doesn't match responses to request. Therefore it can't know when to send a retransmission. I think the request-response should be implemented in another layer above KDSoapUdpClient.

> 
> Other than that itooks lovely.

I added my changes in separate commits for review-ability. I will squash the commits later.

caspermeijn commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,111 @@
+
+KDSoapUdpClient::KDSoapUdpClient(QObject *parent) : 
+    QObject(parent),
+    d_ptr(new KDSoapUdpClientPrivate(this))
```
Fixed

caspermeijn commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,111 @@
+    d_ptr(new KDSoapUdpClientPrivate(this))
+{
+    Q_D(KDSoapUdpClient);
+    d->socketIPv4 = new QUdpSocket(this);
```
I believe it isn't leaking, as the socket has a parent.

caspermeijn commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,111 @@
+                    anySuccess = (writtenSize == data.size()) || anySuccess;
+                } else if(address.protocol() == QAbstractSocket::IPv6Protocol) {
+                    d->socketIPv6->setMulticastInterface(iface);
+                    qint64 writtenSize = d->socketIPv6->writeDatagram(QNetworkDatagram(data, address, port));
```
The Qt documentation is pretty sparse. However, it worked on my laptop...

apachelogger replied:
Mh, I tried my three million interfaces and it works indeed :ok_hand:

caspermeijn commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,111 @@
+bool KDSoapUdpClient::sendMessage(const KDSoapMessage &message, const KDSoapHeaders &headers, const QHostAddress& address, quint16 port) {
+    Q_D(KDSoapUdpClient);
+    KDSoapMessageWriter msgWriter;
+    msgWriter.setVersion(KDSoap::SOAP1_2);
```
Added for sending messages. However this is more difficult for receiving. Should I just assume that the response will be in the same SOAP version as the request or do I need to detect the SOAP version?

apachelogger replied:
I honestly don't know.
KDSoapPendingCall however seems to assume that the reply will be the same version, which does make sense to me anyway.

apachelogger commented:
> I was able to do this for the sending test. However I have to add extra code to make is happens receiving code. Do you think that is worth the trouble?

Don't think so, should it become a problem in the future one can always put in the extra work.

apachelogger commented:
looks good to me :+1:

apachelogger commented:
ping (:

dfaure-kdab commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,133 @@
+void KDSoapUdpClientPrivate::recievedDatagram(const QByteArray &messageData, const QHostAddress &senderAddress, quint16 senderPort)
+{
+    Q_Q(KDSoapUdpClient);
+    //qDebug() << "Recieved datagram from:" << senderAddress << "data:" << QString::fromUtf8(messageData);
```
typo: Received

caspermeijn replied:
Fixed, also in function name

dfaure-kdab commented on unittests/soap_over_udp/test_soap_over_udp.cpp:
```diff
@@ -0,0 +1,173 @@
+        QVERIFY(spy.wait(1000));
+        QList<QVariant> arguments = spy.takeFirst();
+        QVERIFY(KDSoapUnitTestHelpers::xmlBufferCompare(arguments.at(0).value<KDSoapMessage>().toXml(), exampleMessage().toXml()));
+        QVERIFY(arguments.at(1).value<KDSoapHeaders>().size() == 0);
```
QCOMPARE

dfaure-kdab commented on unittests/soap_over_udp/test_soap_over_udp.cpp:
```diff
@@ -0,0 +1,173 @@
+    void testSendMessage() {
+        QUdpSocket testSocket;
+        bool rc = testSocket.bind();
+        Q_ASSERT(rc);
```
QVERIFY

dfaure-kdab commented on unittests/soap_over_udp/test_soap_over_udp.cpp:
```diff
@@ -0,0 +1,173 @@
+        
+        auto data = exampleTextData();
+        qint64 size = testSocket.writeDatagram(data, QHostAddress::LocalHost, 12345);
+        QVERIFY(size == data.size());
```
QCOMPARE (so that we can see values in case of failure)

caspermeijn replied:
Makes sense

caspermeijn commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,133 @@
+
+static const QHostAddress AnyIPv4(QLatin1String("0.0.0.0"));
+
+bool isMulticastAddress(const QHostAddress &address) {
```
Good point

caspermeijn commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,133 @@
+    
+    if (isMulticastAddress(address)) {
+        bool anySuccess = false;
+        for (const auto &iface : QNetworkInterface::allInterfaces()) {
```
Done

anoymouserver commented on src/KDSoapClient/KDSoapUdpClient_p.h:
```diff
@@ -0,0 +1,47 @@
+
+#include <QObject>
+#include <QUdpSocket>
+
```
I had to include `KDSoapValue.h` (used for `KDSoap::SoapVersion` in line 40) and forward-declare `KDSoapUdpClient` to build it.

apachelogger commented:
qmake support in #192

caspermeijn commented on src/KDSoapClient/KDSoapUdpClient.cpp:
```diff
@@ -0,0 +1,134 @@
+    bool rc = true;
+    // Workaround for lack of dual stack sockets in Qt4
+    // Qt5 supports binding to QHostAddress::Any, which will listen on both IPv4 and IPv6 interfaces.
+    rc = d->socketIPv4->bind(AnyIPv4, port, mode) && rc;
```
I am surprised that the code was not compatible with Qt4. I did build this patch using Qt4 to make sure it works. Maybe I build an earlier version and broke the build after that.

In your commit fixing the Qt4 build you mention "once we drop Qt4 support". Are there any plans for dropping Qt4 support? In what time frame could that be expected? Would it be acceptable to make this feature Qt5 only? 

The recent PKCS12 feature only works with Qt5.4, because of the API improvements in that version. But this feature should be able to work correctly using Qt4.

Personally I don't use Qt4 and I can't easily test using Qt4 either (as you have noticed by me breaking Qt4 builds several times ;-). I would happily modify KDSoapUdpClient to be Qt5 only and remove the workaround.

dfaure-kdab commented on unittests/soap_over_udp/test_soap_over_udp.cpp:
```diff
@@ -0,0 +1,173 @@
+        
+        udpClient.sendMessage(exampleMessage(), KDSoapHeaders(), QHostAddress::LocalHost, testSocket.localPort());
+        
+        QVERIFY(testSocket.hasPendingDatagrams());
```
This test just failed on OSX:

********* Start testing of SoapOverUdpTest *********
Config: Using QtTest library 5.8.0, Qt 5.8.0 (x86_64-little_endian-lp64 shared (dynamic) release build; by Clang 7.0.2 (clang-700.1.81) (Apple))
PASS   : SoapOverUdpTest::initTestCase()
PASS   : SoapOverUdpTest::testExample()
FAIL!  : SoapOverUdpTest::testSendMessage() 'testSocket.hasPendingDatagrams()' returned FALSE. ()
   Loc: [/Users/build/Buildbot/kdsoap-qt5.8-osx2/source/unittests/soap_over_udp/test_soap_over_udp.cpp(94)]
PASS   : SoapOverUdpTest::testReceiveMessage()
PASS   : SoapOverUdpTest::cleanupTestCase()
Totals: 4 passed, 1 failed, 0 skipped, 0 blacklisted, 5ms
********* Finished testing of SoapOverUdpTest *********

Maybe this should use QTRY_VERIFY? I can see how the OS does not guarantee that it's immediately available.

caspermeijn replied:
Sounds reasonable. This was one of the reasons that I added to verify, to ensure the message has arrived before proceeding.

caspermeijn replied:
Do you want me to create this change and the change for disabling KDSoapUdpClient for Qt4?

dfaure-kdab replied:
I'll test the change and I'll come back to you if it still fails.

The unittest is already disabled for Qt4, and we're dropping Qt4 support after the upcoming release, so that's fine.
